scripts/gen_vimdoc.py

```diff
@@ -835,33 +886,156 @@ def is_program_listing(para):
+
+    annotations: List[str]
+    """Attributes, e.g., FUNC_API_REMOTE_ONLY. See annotation_map"""
+
+    notes: List[Docstring]
+    """Notes: (@note strings)"""
+
+    signature: str
+    """Function signature with *tags*."""
+
+    parameters: List[FunctionParam]
+    """Parameters: (type, name)"""
+
+    parameters_doc: Dict[str, Docstring]
+    """Parameters documentation. Key is parameter name, value is doc."""
+
+    doc: List[Docstring]
+    """Main description for the function. Separated by paragraph."""
+
+    return_: List[Docstring]
+    """Return:, or Return (multiple): (@return strings)"""
+
+    seealso: List[Docstring]
+    """See also: (@see strings)"""
+
+    xrefs: List[Docstring]
+    """XRefs. Currently only used to track Deprecated functions."""
+
+    # for INCLUDE_C_DECL
+    c_decl: str | None = None
+
+    prerelease: bool = False
+
+    def export_mpack(self) -> Dict[str, Any]:
+        """Convert a dict to be exported as mpack data."""
+        exported = self.__dict__.copy()
+        del exported['notes']
+        del exported['c_decl']
+        del exported['prerelease']
+        del exported['xrefs']
+        exported['return'] = exported.pop('return_')
+        return exported
+
+    def doc_concatenated(self) -> str:
+        """Concatenate all the paragraphs in `doc` into a single string, but
+        remove blank lines before 'programlisting' blocks. #25127
+
+        BEFORE (without programlisting processing):
+            ```vimdoc
+            Example:
+
+            >vim
+                :echo nvim_get_color_by_name("Pink")
+            <
+            ```
+
+        AFTER:
+            ```vimdoc
+            Example: >vim
+                :echo nvim_get_color_by_name("Pink")
+            <
+            ```
+        """
+        def is_program_listing(paragraph: str) -> bool:
+            lines = paragraph.strip().split('\n')
+            return lines[0].startswith('>') and lines[-1] == '<'
+
+        rendered = []
+        for paragraph in self.doc:
+            if is_program_listing(paragraph):
+                rendered.append(' ')  # Example: >vim
+            elif rendered:
+                rendered.append('\n\n')
+            rendered.append(paragraph)
+        return ''.join(rendered)
+
+    def render(self) -> Docstring:
+        """Renders function documentation as Vim :help text."""
+        rendered_blocks: List[Docstring] = []
+
+        def fmt_param_doc(m):
+            """Renders a params map as Vim :help text."""
+            max_name_len = max_name(m.keys()) + 4
+            out = ''
+            for name, desc in m.items():
+                if name == 'self':
+                    continue
+                name = '  • {}'.format('{{{}}}'.format(name).ljust(max_name_len))
+                out += '{}{}\n'.format(name, desc)
+            return out.rstrip()
+
+        # Generate text from the gathered items.
+        chunks: List[Docstring] = [self.doc_concatenated()]
```

Comment:
Is an explicit type annotation needed here? And should the return type of `doc_concatenated` be `Docstring` instead?

Reply:
Fixed typing on `doc_concatenated`. Whenever applicable, it's my personal preference to write type annotations (unless it's verbose and formats ugly) because it gives me more hints.